src/use-queue.ts

```diff
@@ -0,0 +1,50 @@
+import { useState, useMemo } from 'react'
+import produce from 'immer'
+
+export const useQueue = <T>(initialValue: T[] = []) => {
+  const [value, setValue] = useState(initialValue)
+
+  const handlers = useMemo(
+    () => ({
+      dequeue: () => {
+        if (value.length === 0) return
+
+        let shifted
+        setValue((oldQueue) =>
+          produce(oldQueue, (draft: T[]) => {
+            shifted = draft.shift()
+          })
+        )
+
+        return shifted
+      },
+
+      enqueue: (item: T) => {
+        setValue((oldQueue) =>
+          produce(oldQueue, (draft: T[]) => {
+            draft.push(item)
```

Comment:
I'd prefer if we use something like `.unshift` to add it at the start of the array

Reply:
Sure